diff --git a/src/com/android/camera/FocusOverlayManager.java b/src/com/android/camera/FocusOverlayManager.java
index c16d01b..79e8b38 100644
--- a/src/com/android/camera/FocusOverlayManager.java
+++ b/src/com/android/camera/FocusOverlayManager.java
@@ -358,7 +358,7 @@ public class FocusOverlayManager implements PreviewStatusListener.PreviewAreaCha
     }
 
     public void onSingleTapUp(int x, int y) {
-        if (!mInitialized || mState == STATE_FOCUSING_SNAP_ON_FINISH) {
+        if (!mInitialized || mState == STATE_FOCUSING_SNAP_ON_FINISH || mState == STATE_FOCUSING) {
             return;
         }
 
